### ISSUE TITLE
[BUG] backward of solve_sample value does no populate gradients for weights a and b

### ISSUE THREAD

**Issue by rflamary** [jeanfeydy/geomloss#93]
backward on OTResultSample.value populates gradients wrt the samples bu treats the weights as detached.

```python

import geomloss.ot as got

n = 100
rng = np.random.RandomState(0)

x = rng.randn(n, 2)
x2 = rng.randn(n//2, 2)+5

xb = torch.tensor(x, dtype=torch.float32, requires_grad=True)
xb2 = torch.tensor(x2, dtype=torch.float32)

a = torch.ones(n, dtype=torch.float32, requires_grad=True) / n
b = torch.ones(n//2, dtype=torch.float32, requires_grad=True) / (n//2)

sol = got.solve_sample(xb, xb2, a=a, b=b, reg=1, max_iter=1000)

loss = sol.value

loss.backward()

print(xb.grad)
print(a.grad)
print(b.grad)

```

outputs:

```
tensor([[-0.1847, -0.1929],
        [-0.1840, -0.1498],
        [-0.1953, -0.1972],
        [-0.1815, -0.2046],
...
        [-0.1859, -0.1861],
        [-0.1731, -0.1669]])
None
None
UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more information. 
UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more information.
```
